Microsoft.Dynamics365.UIAutomation.Browser/Extensions/SeleniumExtensions.cs

```diff
@@ -547,74 +555,56 @@ public static object WaitForScript(this IWebDriver driver, string script, TimeSp
 
             return null;
         }
-
-        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by)
-        {
-            return WaitUntilAvailable(driver, by, Constants.DefaultTimeout, null, null);
-        }
-
-        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan timeout)
-        {
-            return WaitUntilAvailable(driver, by, timeout, null, null);
-        }
-
+        
         public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, string exceptionMessage)
         {
-            return WaitUntilAvailable(driver, by, Constants.DefaultTimeout, null, d =>
-            {
-                throw new InvalidOperationException(exceptionMessage);
-            });
+            return WaitUntilAvailable(driver, by, Constants.DefaultTimeout, null, d => throw new InvalidOperationException(exceptionMessage));
         }
-
+        
         public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan timeout, string exceptionMessage)
         {
-            return WaitUntilAvailable(driver, by, timeout, null, d =>
-            {
-                throw new InvalidOperationException(exceptionMessage);
-            });
+            return WaitUntilAvailable(driver, by, timeout, null, d => throw new InvalidOperationException(exceptionMessage));
         }
 
-        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan timeout, Action<IWebDriver> successCallback)
+        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan? timeout = null, Action<IWebDriver> successCallback = null, Action<IWebDriver> failureCallback = null)
         {
-            return WaitUntilAvailable(driver, by, timeout, successCallback, null);
+            Func<IWebDriver, IWebElement> conditions = d =>
+            {
+                ReadOnlyCollection<IWebElement> elements = d.FindElements(by);
+                int? count = elements?.Count;
+                if (count == null || count == 0)
+                    return null;
+
+                var result = count > 1
+                    ? elements.FirstOrDefault(x => x?.Displayed == true)
+                    : elements.First(x => x != null);
+                
+                return result;
+            };
+            return WaitUntil(driver, conditions, timeout ?? Constants.DefaultTimeout, successCallback, failureCallback);
         }
 
-        public static IWebElement WaitUntilAvailable(this IWebDriver driver, By by, TimeSpan timeout, Action<IWebDriver> successCallback, Action<IWebDriver> failureCallback)
+        private static IWebElement WaitUntil(IWebDriver driver, Func<IWebDriver, IWebElement> conditions, TimeSpan? timeout, Action<IWebDriver> successCallback, Action<IWebDriver> failureCallback)
```

Comment:
can you make it public. we can use it to implement WaitForNotXXXX when specific methods are required

Reply:
thanks. This PR is big enough, I plan to change/extend the Selenium Extentions in other PR. In this case I was just fixing a bug.

Reply:
Hi @easyrepro. You can check the SeleniumExtensions.cs in #747 the Wait-Methods are now extended. There is also two WaitUntil that you can extend as you want & also a RepeatUntil. I hope did you like it. BR